### PR TITLE
feat: ice connection stats

### PR DESCRIPTION
**Pull request by andrei-gavrilescu**
Adds reconnects, ice failure and session duration.

### CODE REVIEW

File: src/features/quality-stats/StatsAggregator.js [hunk 1]
```diff
@@ -228,22 +270,30 @@ class StatsAggregator {
...
+            pcResults.iceReconnects = this._calculateReconnects(pcData);
+            pcResults.pcSessionDurationMs = this._calculateSessionDurationMs(pcData);
+            pcResults.iceFailed = this._didIceConnectionFail(pcData);
 
-            resultMap[pc].tracks = pcTrackStats;
-            resultMap[pc].trackAggregates = pcTrackResults;
-            resultMap[pc].transportAggregates = pcTransportResults;
+            pcResults.tracks = pcTrackStats;
+            pcResults.trackAggregates = pcTrackResults;
+            pcResults.transportAggregates = pcTransportResults;
```
Comment (nils-ohlmeier) [inline src/features/quality-stats/StatsAggregator.js:294]: Wouldn't it make sense to move these three up to where we call the _calculate...() functions, like you do now for duration and the ICE states?

File: src/test/client.js [hunk 2]
```diff
@@ -370,42 +375,42 @@ function runTest() {
 
     simulateConnection(
         './src/test/dumps/google-standard-stats-p2p',
-        './src/test/results/google-standard-stats-p2p-result.json',
+        './src/test/jest/results/google-standard-stats-p2p-result.json',
```
Comment (nils-ohlmeier) [inline src/test/client.js:378]: Do we want to switch these test over to use one common results set?
Or should we abandon these tests here and switch to only jest?

**Review by nils-ohlmeier — APPROVED** [review]
In general looks good, only a few questions, but no blocker.

File: src/test/client.js (hunk 2 above)
Comment (andrei-gavrilescu) [reply src/test/client.js:378]: So this change switches them to the common result set, however the purpose for this test suite was to be more end to end. In a future refactor I'd suggest we remove testing processing results all together here (maybe keep one or two)  and do some more advanced end to end testing, like running multiple connections at the same time, failing the connection, testing the reconnect mechanism once we add and so on, wdyt?

File: src/features/quality-stats/QualityStatsCollector.js [hunk 3]
```diff
@@ -205,6 +209,48 @@ class QualityStatsCollector {
...
+     *
+     * @param {*} dumpLineObj
+     */
+    processConnectionState(dumpLineObj) {
+        const [ , pc, state, timestamp ] = dumpLineObj;
+
+        const pcData = this._getPcData(pc);
+
+        if (isIceConnected(state) && !pcData.startTime) {
+            pcData.startTime = timestamp;
```
[the remark this answers is not in the document]
Comment (andrei-gavrilescu) [reply src/features/quality-stats/QualityStatsCollector.js:247]: This monitors the connectionstate changes, I might be wrong but afaik it's supposed to be a more complete view of the connection state i.e. ice + dtls, as opposed to iceconnectionstate.
Even so the isIceConnected function name is misleading there, I'll change to something more appropriate.

File: src/features/quality-stats/StatsAggregator.js (hunk 1 above)
Comment (andrei-gavrilescu) [reply src/features/quality-stats/StatsAggregator.js:294]: yes it would :) , changing.

**Review by gpolitis — APPROVED** [review]

**Pull request merged**